gsy_framework/sim_results/carbon_emissions.py

```diff
@@ -0,0 +1,246 @@
+from typing import Union, Dict, List
+
+import requests
+import pandas as pd
+from pendulum import duration
+from entsoe import EntsoePandasClient
+from entsoe.mappings import lookup_area, Area
+from entsoe.parsers import parse_generation
+from geopy.geocoders import Nominatim
+
+from gsy_framework.sim_results.results_abc import ResultsBaseClass
+
+ENTSOE_URL = "https://web-api.tp.entsoe.eu/api"
+
+GENERATION_PLANT_TO_CARBON_EMISSIONS = {
+    # source: https://www.ipcc.ch/site/assets/uploads/2018/02/ipcc_wg3_ar5_annex-iii.pdf#page=7
+    # keys match the generation plants in the entsoe API
+    # values in gCO2eq/kWh
+    "Fossil Gas": {"min": 410, "median": 490, "max": 650},
+    "Wind Offshore": {"min": 8.0, "median": 12, "max": 35},
+    "Hydro Pumped Storage": {"min": 1.0, "median": 24, "max": 2200},
+    "Nuclear": {"min": 3.7, "median": 12, "max": 110},
+    "Biomass": {"min": 620, "median": 740, "max": 890},
+}
+
+GEOPY_CONTRY_CODE_TO_ENTSOE_CODE = {
+    "BE": "BE",
+    "PT": "PT",
+    "ES": "ES",
+    "FR": "FR",
+    "DE": "DE",
+    "NL": "NL",
+    "IT": "IT",
+    "SE": "SE",
+    "NO": "NO",
+    "FI": "FI",
+    "DK": "DK",
+    "AT": "AT",
+    "CH": "CH",
+    "GB": "GB",
+    "IE": "IE",
+    "PL": "PL",
+    "CZ": "CZ",
+    "SK": "SK",
+    "HU": "HU",
+    "SI": "SI",
+    "HR": "HR",
+    "RO": "RO",
+    "BG": "BG",
+    "GR": "GR",
+    "EE": "EE",
+    "LV": "LV",
+    "LT": "LT",
+    "LU": "LU",
+    "MT": "MT",
+    "CY": "CY",
+}
+
+
+class CarbonEmissionsHandler(EntsoePandasClient, ResultsBaseClass):
```

Comment:
I do not particularly like the multiple inheritance here, since the same outcome can be achieved by composition. I think it will be much more readable if you create a separate EntsoeClient class that inherits from EntsoePandasClient, override whatever you want and construct it in the CarbonEmissionsHandler constructor. That way you will avoid multiple inheritance, and the dependencies between the Entsoe client and our logic will be far more decoupled.

Reply:
Thanks for the suggestion. This was a really good catch. I replaced the multiple inheritance in with composition through the EntsoePandasClientAdapter